libafl/src/inputs/bytessub.rs

```diff
@@ -27,14 +27,112 @@ fn end_index<R>(range: &R, max_len: usize) -> usize
 where
     R: RangeBounds<usize>,
 {
-    match range.end_bound() {
+    let end = match range.end_bound() {
         Bound::Unbounded => max_len,
         Bound::Included(end) => end + 1,
         Bound::Excluded(end) => *end,
+    };
+
+    min(end, max_len)
+}
+
+/// An immutable contiguous subslice of an input implementing [`HasTargetBytes`].
+/// It is mostly useful to cheaply wrap a subslice of a given input.
+///
+/// A mutable version is available: [`BytesSubInputMut`].
+#[derive(Debug)]
+pub struct BytesSubInput<'a> {
+    /// The (complete) parent input we will work on
+    parent_input: OwnedSlice<'a, u8>,
+    /// The range inside the parent input we will work on
+    range: Range<usize>,
+}
+
+impl<'a> BytesSubInput<'a> {
+    /// Creates a new [`BytesSubInputMut`] that's a view on an input with mutator bytes.
+    /// The sub input can then be used to mutate parts of the original input.
+    pub fn new<I, R>(parent_input: &'a I, range: R) -> Self
```

Comment:
Instead of taking an input here, just take a slice?

Reply:
ok i think it would make more sense indeed, I'll give it a try